Compiler throws an exception when unicode escape is used in "raw"-interpolated multiline string (when printing --deprecation warning)

Scala 2.13.18
```scala
lazy val scala2Module = project.settings(
  scalaVersion := "2.13.18",
  scalacOptions ++= Seq("--deprecation")
)
```

```scala
object Scala2Module {
  val singleLine = "string\u0023"
  val singleLineRaw = raw"string\u0023" //Compiler warning: Unicode escapes in raw interpolations are deprecated; use literal characters instead
  val multiline = """string\u0023""" //Compile warning: Unicode escapes in triple quoted strings are deprecated; use the literal character instead
  val multilineRaw = raw"""string\u0023"""
}
```

# Expected result
Compiler should produce a warning
> Unicode escapes in triple quoted strings are deprecated; use the literal character instead

for  this code
```
raw"""string\u0023"""
```



# Actual result
Compiler throws an exception
```
scala: ## Exception when compiling 1 sources to ...
java.lang.IndexOutOfBoundsException: 396
scala.reflect.internal.util.BatchSourceFile.offsetToLine(SourceFile.scala:213)
scala.tools.xsbt.DelegatingReporter$.makePosition$1(DelegatingReporter.scala:137)
scala.tools.xsbt.DelegatingReporter$.convert(DelegatingReporter.scala:190)
scala.tools.xsbt.DelegatingReporter.doReport(DelegatingReporter.scala:225)
scala.reflect.internal.Reporter.filteredInfo(Reporting.scala:126)
scala.reflect.internal.Reporter.warning(Reporting.scala:118)
scala.tools.nsc.Reporting$PerRunReporting.scala$tools$nsc$Reporting$PerRunReporting$$issueWarning(Reporting.scala:229)
scala.tools.nsc.Reporting$PerRunReporting.issueIfNotSuppressed(Reporting.scala:248)
scala.tools.nsc.Reporting$PerRunReporting.$anonfun$reportSuspendedMessages$4(Reporting.scala:191)
scala.tools.nsc.Reporting$PerRunReporting.$anonfun$reportSuspendedMessages$4$adapted(Reporting.scala:191)
scala.collection.mutable.LinkedHashSet.foreach(LinkedHashSet.scala:128)
scala.tools.nsc.Reporting$PerRunReporting.$anonfun$reportSuspendedMessages$3(Reporting.scala:191)
scala.tools.nsc.Reporting$PerRunReporting.reportSuspendedMessages(Reporting.scala:191)
scala.tools.nsc.typechecker.Analyzer$typerFactory$TyperPhase.apply(Analyzer.scala:159)
scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:485)
scala.tools.nsc.typechecker.Analyzer$typerFactory$TyperPhase.run(Analyzer.scala:130)
scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1559)
scala.tools.nsc.Global$Run.compileUnits(Global.scala:1543)
scala.tools.nsc.Global$Run.compileSources(Global.scala:1535)
scala.tools.nsc.Global$Run.compileFiles(Global.scala:1651)
scala.tools.xsbt.CachedCompiler0.run(CompilerBridge.scala:176)
scala.tools.xsbt.CachedCompiler0.run(CompilerBridge.scala:139)
scala.tools.xsbt.CompilerBridge.run(CompilerBridge.scala:43)
sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:196)
scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:252)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:186)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:166)
sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:241)
sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:166)
sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:214)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:542)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:542)
sbt.internal.inc.Incremental$.$anonfun$apply$3(Incremental.scala:182)
sbt.internal.inc.Incremental$.$anonfun$apply$3$adapted(Incremental.scala:180)
sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:458)
sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:264)
sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:413)
sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:500)
sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:400)
sbt.internal.inc.Incremental$.apply(Incremental.scala:208)
sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:542)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:496)
sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:433)
sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
org.jetbrains.jps.incremental.scala.local.SbtCompiler.$anonfun$doCompile$3(SbtCompiler.scala:87)
scala.util.Try$.apply(Try.scala:217)
org.jetbrains.jps.incremental.scala.local.SbtCompiler.doCompile(SbtCompiler.scala:85)
org.jetbrains.jps.incremental.scala.local.SbtCompiler.compile(SbtCompiler.scala:17)
org.jetbrains.jps.incremental.scala.local.LocalServer.doCompile(LocalServer.scala:47)
org.jetbrains.jps.incremental.scala.local.LocalServer.compile(LocalServer.scala:26)
org.jetbrains.jps.incremental.scala.remote.Main$.compileLogic(Main.scala:213)
org.jetbrains.jps.incremental.scala.remote.Main$.$anonfun$handleCommand$1(Main.scala:194)
org.jetbrains.jps.incremental.scala.remote.Main$.decorated$1(Main.scala:181)
org.jetbrains.jps.incremental.scala.remote.Main$.handleCommand(Main.scala:191)
org.jetbrains.jps.incremental.scala.remote.Main$.serverLogic(Main.scala:164)
org.jetbrains.jps.incremental.scala.remote.Main$.nailMain(Main.scala:104)
org.jetbrains.jps.incremental.scala.remote.Main.nailMain(Main.scala)
java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
java.base/java.lang.reflect.Method.invoke(Method.java:580)
com.facebook.nailgun.NGSession.runImpl(NGSession.java:312)
com.facebook.nailgun.NGSession.run(NGSession.java:198)
```

Thanks I'll try again. I assumed a position problem.

Oh wait, scalac only does single carets. But you showed scala-cli.

Obviously, I did something different. Or maybe you're doing something different.

```
➜  scala git:(issue/13161-raw-warn-eof) scala-cli compile --server=false -S 2.13.18 test/files/neg/t13161.scala
/home/amarki/projects/scala/test/files/neg/t13161.scala:6: warning: Unicode escapes in triple quoted strings are deprecated; use the literal character instead
  val multiline = """string\u0023""" //Compile warning: Unicode escapes in triple quoted strings are deprecated; use the literal character instead
                           ^
/home/amarki/projects/scala/test/files/neg/t13161.scala:5: warning: Unicode escapes in raw interpolations are deprecated; use literal characters instead
  val singleLineRaw = raw"string\u0023" //Compiler warning: Unicode escapes in raw interpolations are deprecated; use literal characters instead
                                ^
/home/amarki/projects/scala/test/files/neg/t13161.scala:7: warning: Unicode escapes in raw interpolations are deprecated; use literal characters instead
  val multilineRaw = raw"""string\u0023"""
                                 ^
error: No warnings can be incurred under -Werror.
3 warnings
1 error
Compilation failed
```
Same result as copy/paste from your cli example.

--- a/src/compiler/scala/tools/reflect/FastStringInterpolator.scala
+++ b/src/compiler/scala/tools/reflect/FastStringInterpolator.scala
@@ -41,23 +41,24 @@ trait FastStringInterpolator extends FormatInterpolator {
       }
       val treated = parts.mapConserve {
         case lit @ Literal(Constant(stringVal: String)) =>
-          def asRaw = if (currentRun.sourceFeatures.unicodeEscapesRaw) stringVal else {
-            val processed = processUnicode(stringVal)
-            if (processed == stringVal) stringVal else {
-              val pos = {
-                val diffindex = processed.zip(stringVal).zipWithIndex.collectFirst {
-                  case ((p, o), i) if p != o => i
-                }.getOrElse(processed.length - 1)
-                lit.pos.withShift(diffindex)
-              }
-              def msg(fate: String) = s"Unicode escapes in raw interpolations are $fate; use literal characters instead"
-              if (currentRun.isScala3)
-                runReporting.warning(pos, msg("ignored in Scala 3 (or with -Xsource-features:unicode-escapes-raw)"), Scala3Migration, c.internal.enclosingOwner)
-              else
-                runReporting.deprecationWarning(pos, msg("deprecated"), "2.13.2", "", "")
-              processed
+          def asRaw = if (currentRun.sourceFeatures.unicodeEscapesRaw) stringVal else
+            processUnicode(stringVal) match {
+              case `stringVal` => stringVal
+              case processed =>
+                val pos = {
+                  val index = processed.zip(stringVal).indexWhere { case (p, s) => p != s }
+                  adjustedEscPos(lit.pos, index)
+                }
+                def msg(fate: String) =
+                  s"Unicode escapes in raw interpolations are $fate; use literal characters instead"
+                def ignored =
+                  msg("ignored in Scala 3 (or with -Xsource-features:unicode-escapes-raw)")
+                if (currentRun.isScala3)
+                  runReporting.warning(pos, msg=ignored, Scala3Migration, site=c.internal.enclosingOwner)
+                else
+                  runReporting.deprecationWarning(pos, msg=msg("deprecated"), since="2.13.2", site="", origin="")
+                processed
             }
-          }
           val value =
             try if (isRaw) asRaw else processEscapes(stringVal)
             catch {
@@ -65,7 +66,8 @@ trait FastStringInterpolator extends FormatInterpolator {
               case iue: InvalidUnicodeEscapeException => c.abort(adjustedEscPos(lit.pos, iue.index), iue.getMessage)
             }
           val k = Constant(value)
-          // To avoid the backlash of backslash, taken literally by Literal, escapes are processed strictly (scala/bug#11196)
+          // To avoid the backlash of backslash, taken literally by Literal, escapes are processed strictly.
+          // scala/bug#11196
           treeCopy.Literal(lit, k).setType(ConstantType(k))
         case x => throw new MatchError(x)
       }

--- a/test/files/neg/t13161.check
+++ b/test/files/neg/t13161.check
@@ -1,0 +1,6 @@
+[WARNING] [source-t13161.scala,line-8,offset=346]: Unicode escapes in triple quoted strings are deprecated; use the literal character instead
+[WARNING] [RangePosition(t13161.scala, 206, 206, 208)]: Unicode escapes in raw interpolations are deprecated; use literal characters instead
+[WARNING] [RangePosition(t13161.scala, 499, 499, 501)]: Unicode escapes in raw interpolations are deprecated; use literal characters instead
+[ERROR] [NoPosition]: No warnings can be incurred under -Werror.
+3 warnings
+1 error

--- a/test/files/neg/t13161.scala
+++ b/test/files/neg/t13161.scala
@@ -1,0 +1,10 @@
+//> using options -deprecation -Werror -Xreporter:scala.tools.partest.nest.PlainReporter
+//
+// todo -Yvalidate-pos:_
+
+object Scala2Module {
+  val singleLine = "string\u0023"
+  val singleLineRaw = raw"string\u0023" //Compiler warning: Unicode escapes in raw interpolations are deprecated; use literal characters instead
+  val multiline = """string\u0023""" //Compile warning: Unicode escapes in triple quoted strings are deprecated; use the literal character instead
+  val multilineRaw = raw"""string\u0023"""
+}

